### PR TITLE
[JENKINS-61840] Restyle buttons

### PR DESCRIPTION
**Pull request by fqueiruga**
See [JENKINS-61840](https://issues.jenkins-ci.org/browse/JENKINS-61840).

This PR can be tested using the following docker image:
`docker run --rm -ti -p 8080:8080 -e ID=4658 jenkins/core-pr-tester`

This PR restyles the buttons. The CSS API is described in detail in the JIRA issue.

To summarize:
- Buttons are restyled from scratch. Removed references to YahooUI CSS code that dealt with buttons in order to avoid overriding styles.
- Use the same CSS API that YUI buttons used. No breaking changes here.
- Created styles for 4 button variants: primary, secondary (default), danger and link (transparent). Added a _large_ version as well.
- Added support for buttons with leading and trailing buttons.
- Added support for icon-only buttons.

Screenshot:
Captura de pantalla 2020-04-15 a las 13 36 26

<details>
<summary>Sample jelly code to test the buttons</summary>

````
    <div style="margin-top: 2rem; display: flex; flex-wrap: wrap">
      <strong style="display:block; width: 100%; text-align: left; margin-bottom: .5rem">Anchor buttons</strong>

      <a href="#" class="yui-button primary">Primary</a>
      <a href="#" class="yui-button" >Default</a>
      <a href="#" class="yui-button danger">Danger</a>
      <a href="#" class="yui-button link-button">Link</a>
    </div>

    <div style="margin-top: 2rem; display: flex; flex-wrap: wrap">
      <strong style="display:block; width: 100%; text-align: left; margin-bottom: .5rem">Large anchor buttons (disabled)</strong>

      <a href="#" class="yui-button large-button disabled primary">Primary</a>
      <a href="#" class="yui-button large-button disabled">Default</a>
      <a href="#" class="yui-button large-button disabled danger">Danger</a>
      <a href="#" class="yui-button large-button disabled link-button">Link</a>
    </div>

    <div style="margin-top: 2rem; display: flex; flex-wrap: wrap">
      <strong style="display:block; width: 100%; text-align: left; margin-bottom: .5rem">Buttons with leading icons</strong>

      <span class="yui-button primary">
        <button>
          <l:svgIcon href="${resURL}/images/material-icons/svg-sprite-action-symbol.svg#ic_done_24px" class="leading-icon" />
          <span>Primary</span>
        </button>
      </span>
      <span class="yui-button">
        <button>
          <l:svgIcon href="${resURL}/images/material-icons/svg-sprite-action-symbol.svg#ic_cached_24px" class="leading-icon" />
          <span>Default</span>
        </button>
      </span>
      <span class="yui-button danger">
        <button>
          <l:svgIcon href="${resURL}/images/material-icons/svg-sprite-action-symbol.svg#ic_report_problem_24px" class="leading-icon" />
          <span>Danger</span>
        </button>
      </span>
      <span class="yui-button link-button">
        <button>
          <l:svgIcon href="${resURL}/images/material-icons/svg-sprite-action-symbol.svg#ic_help_24px" class="leading-icon" />
          <span>Link</span>
        </button>
      </span>
    </div>

    <div style="margin-top: 2rem; display: flex; flex-wrap: wrap">
      <strong style="display:block; width: 100%; text-align: left; margin-bottom: .5rem">Large buttons with leading and trailing icons</strong>

      <span class="yui-button large-button primary">
        <button>
          <l:svgIcon href="${resURL}/images/material-icons/svg-sprite-action-symbol.svg#ic_done_24px" class="leading-icon" />
          <span>Primary</span>
        </button>
      </span>
      <span class="yui-button large-button">
        <button>
          <l:svgIcon href="${resURL}/images/material-icons/svg-sprite-action-symbol.svg#ic_cached_24px" class="leading-icon" />
          <span>Default</span>
        </button>
      </span>
      <span class="yui-button large-button danger">
        <button>
          <span>Danger</span>
          <l:svgIcon href="${resURL}/images/material-icons/svg-sprite-action-symbol.svg#ic_report_problem_24px" class="trailing-icon" />
        </button>
      </span>
      <span class="yui-button large-button link-button">
        <button>
          <span>Link</span>
          <l:svgIcon href="${resURL}/images/material-icons/svg-sprite-action-symbol.svg#ic_open_in_new_24px" class="trailing-icon" />
        </button>
      </span>
    </div>

    <div style="margin-top: 2rem; display: flex; flex-wrap: wrap">
      <strong style="display:block; width: 100%; text-align: left; margin-bottom: .5rem">Normal and large icon buttons</strong>

      <span class="yui-button icon-button">
        <button>
          <l:svgIcon href="${resURL}/images/material-icons/svg-sprite-social-symbol.svg#ic_notifications_24px" />
        </button>
      </span>
      <span class="yui-button primary icon-button">
        <button>
          <l:svgIcon href="${resURL}/images/material-icons/svg-sprite-action-symbol.svg#ic_cached_24px" />
        </button>
      </span>
      <span class="yui-button danger icon-button large-button">
        <button>
          <l:svgIcon href="${resURL}/images/material-icons/svg-sprite-social-symbol.svg#ic_notifications_24px" />
        </button>
      </span>
      <span class="yui-button link-button icon-button large-button">
        <button>
          <l:svgIcon href="${resURL}/images/material-icons/svg-sprite-action-symbol.svg#ic_cached_24px" />
        </button>
      </span>
    </div>
````
</details>

### Proposed changelog entries

* Buttons were restyled. Added support for large buttons, hyperlinks styled as buttons and icon-only buttons.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@timja 
@daniel-beck 
@oleg-nenashev 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a JIRA issue should exist and be labeled as `lts-candidate`

### CODE REVIEW

File: core/src/main/resources/lib/form/apply.jelly [hunk 1]
```diff
@@ -35,8 +35,15 @@ THE SOFTWARE.
     <s:attribute name="value">
       The text of the apply button.
     </s:attribute>
+
+    <s:attribute name="isLarge">
```
Comment (timja) [inline core/src/main/resources/lib/form/apply.jelly:39]:
```suggestion
    <s:attribute name="large">
```

File: core/src/main/resources/lib/form/apply.jelly [hunk 2]
```diff
@@ -35,8 +35,15 @@ THE SOFTWARE.
...
+        Set to "true" to enable the large variant of the button
+    </s:attribute>
   </s:documentation>
   <st:adjunct includes="lib.form.apply.apply"/>
   <input type="hidden" name="core:apply" value="" />
-  <input type="button" value="${attrs.value ?: '%Apply'}" class="apply-button applyButton" name="Apply"/><!-- applyButton is legacy -->
+  <input type="button"
+         value="${attrs.value ?: '%Apply'}"
+         class="apply-button applyButton ${attrs.isLarge ? 'large-button' : ''}"
```
Comment (timja) [inline core/src/main/resources/lib/form/apply.jelly:47]:
```suggestion
         class="apply-button applyButton ${attrs.large ? 'large-button' : ''}"
```

File: core/src/main/resources/hudson/model/Job/configure.jelly [hunk 3]
```diff
@@ -65,8 +65,8 @@ THE SOFTWARE.
             <j:if test="${h.hasPermission(it,it.CONFIGURE)}">
               <f:bottomButtonBar>
                 <!--<input type="button" name="StructureTest" value="Test" onclick="buildFormTree(this.form)" />-->
-                <f:submit value="${%Save}"/>
-                <f:apply value="${%Apply}"/>
+                <f:submit value="${%Save}" isLarge="true"/>
```
Comment (timja) [inline core/src/main/resources/hudson/model/Job/configure.jelly:68]:
```suggestion
                <f:submit value="${%Save}" large="true"/>
```

File: core/src/main/resources/hudson/model/Job/configure.jelly [hunk 4]
```diff
@@ -65,8 +65,8 @@ THE SOFTWARE.
...
+                <f:apply value="${%Apply}" isLarge="true" />
```
Comment (timja) [inline core/src/main/resources/hudson/model/Job/configure.jelly:69]:
```suggestion
                <f:apply value="${%Apply}" large="true" />
```

File: core/src/main/resources/lib/form/submit.jelly [hunk 5]
```diff
@@ -36,6 +36,14 @@ THE SOFTWARE.
     <s:attribute name="value" use="required">
       The text of the submit button. Something like "submit", "OK", etc.
     </s:attribute>
+
+    <s:attribute name="isLarge">
```
Comment (timja) [inline core/src/main/resources/lib/form/submit.jelly:40]:
```suggestion
    <s:attribute name="large">
```

File: core/src/main/resources/lib/form/submit.jelly [hunk 6]
```diff
@@ -36,6 +36,14 @@ THE SOFTWARE.
...
+      Set to "true" to enable the large variant of the button
+    </s:attribute>
   </s:documentation>
-  <input type="submit" name="${attrs.name ?: 'Submit'}" value="${attrs.value ?: '%Submit'}" class="submit-button primary" />
+
+  <input type="submit"
+         name="${attrs.name ?: 'Submit'}"
+         value="${attrs.value ?: '%Submit'}"
+         class="submit-button primary ${attrs.isLarge ? 'large-button' : ''}" />
```
Comment (timja) [inline core/src/main/resources/lib/form/submit.jelly:48]:
```suggestion
         class="submit-button primary ${attrs.large ? 'large-button' : ''}" />
```

**Review by timja — COMMENTED** [review]
From an initial look I would prefer not to call the flag isLarge,

I'll pull this locally and take a look as well

**Review by timja — APPROVED** [review]
the email notification section in /configure looks a bit funny but all else looks good

image

**Comment by fqueiruga** [comment]
I have fixed the icon alignment on the email notification section:
Captura de pantalla 2020-04-15 a las 15 53 14

**Comment by timja** [comment]
nice thanks Felix

**Comment by alecharp** [comment]
I know this might be out of scope, but we are still using `yui` prefix, when those classes, IIRC, came from the Yahoo UI library. As we are not using that anymore, maybe we could rename the classes. WDYT?

**Comment by fqueiruga** [comment]
We could not just rename the classes because the current CSS API is being used across the whole ecosystem. There are also many cases were we will never be able to skip the `.yui-*` classes, such as menu buttons, dropdowns, etc.

Because we need to support those features and the existing code, I avoided coming up with a new interface that would never truly replace the old one, and focused on making sure it worked with existing code. On the new addition to the button catalog, hyperlink buttons, I just kept the `.yui-button` class for consistency.

That said, I'm not 100% happy with needing to wrap buttons within a span tag, so I'm considering adding an alternative. Do you have any concrete vision for what a new CSS API would be @alecharp ?

**Review by romenrg — COMMENTED** [review]
It's great to see more things on the UI being refreshed. Thanks @fqueiruga !

This buttons change seemed a bit subtle to me at first, but after playing a bit with it now, I start to notice the improvement more and more.

One thing I noted, though, is that the "OK" button that appears on the "Create new item" page, does not seem to be updated. So far is the only one I've detected that seems to still show with the old styles.

**Screenshots from my local execution on that page:**

The "Create new item" page and the OK button:
Create New Item


And the over effect in that button:
Create New Item hover

As I said, in the rest of screens I tested, I saw the new styles.

**Review by StefanSpieker — APPROVED** [review]
I really like the new buttons. Thanks for the restyling 👍

**Review by alecharp — APPROVED** [review]
I tested locally and I didn't see any problem. 

@fqueiruga for the renaming, it was just an idea. Normally, plugin devs shouldn't be calling the classes directly, but be using the jelly tags. But you are correct, we never know.

**Comment by fqueiruga** [comment]
I think buttons, hyperlinks, etc  are UI primitives that I would expect everybody could use without repercusions. Not everything can go through jelly if that makes sense.

**Comment by timja** [comment]
> I think buttons, hyperlinks, etc are UI primitives that I would expect everybody could use without repercusions. Not everything can go through jelly if that makes sense.

I wouldn't expect many buttons not to go through jelly, links sure...

**Comment by fqueiruga** [comment]
Fixed that problem Romén, re-requesting your review

**Review by varyvol — APPROVED** [review]

**Review by romenrg — APPROVED** [review]
Awesome @fqueiruga .

Just tested the new changes locally and the "new item" page does indeed show the new styles in the "OK" button now as well. Thanks!

**Comment by timja** [comment]
Looks great thanks Felix and thanks everyone for all the reviews and trying it out

We'll merge tomorrow if no negative feedback

**Comment by daniel-beck** [comment]
Looks great.

The only minor issue I found was that the expandable multi-line input fields now look even weirder than before, but they were really ugly before already, so no big deal.

> Screenshot 2020-04-18 at 00 52 25

**Review by daniel-beck — APPROVED** [review]

File: core/src/main/resources/lib/layout/layout.jelly [hunk 7]
```diff
@@ -153,9 +153,7 @@ THE SOFTWARE.
     </j:if>
 
     <link rel="stylesheet" href="${resURL}/scripts/yui/container/assets/container.css" type="text/css"/>
-    <link rel="stylesheet" href="${resURL}/scripts/yui/assets/skins/sam/skin.css" type="text/css" />
```
Comment (daniel-beck) [inline core/src/main/resources/lib/layout/layout.jelly:156]: This probably caused https://issues.jenkins-ci.org/browse/JENKINS-62001

**Pull request merged**